core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -537,7 +538,8 @@ class ReplicaManager(val config: KafkaConfig,
           val initialLogEndOffset = localReplica.logEndOffset
           val logReadInfo = localReplica.log match {
             case Some(log) =>
-              log.read(offset, fetchSize, maxOffsetOpt)
+              val adjustedFetchSize = if (TopicConstants.INTERNAL_TOPICS.contains(topic) && !readOnlyCommitted) Math.max(fetchSize, log.config.maxMessageSize) else fetchSize
```

Comment:
@ewencp I agree that we should let the consumer know the max message size. In fact I think  max message size as well as some other important topic configuration should be part of the topic metadata. That said, I think the main concern here is about memory management. So basically let's say the consumer only wants to fetch 1MB and has already allocated 1MB of buffer, if we do not honor the max fetch size the consumer specified, this would break them. If we really want to change the behavior of max fetch size, maybe it deserves a KIP.

I am not sure if we should address the scenario of having different max message size configuration on the brokers in the same cluster because it is almost surely a user misconfiguration. 

Another reason we only loosen the internal topic only is because at the end of the day, we just want to store the metadata somewhere, and we happened to choose a Kafka topic as the storage. So the __consumer_offset topic is special because it is more our internal data storage than a topic. So making sure the offset topic can be replicated is different from a normal consumer fetch.